File: single.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class LL:
    def __init__(self):
        self.head = None
        self.size = 0

    def insertAtBeginning(self, data):
        newnode = Node(data)
        newnode.next = self.head
        self.head = newnode
        self.size += 1
        self.display()

    def insertAtEnd(self, data):
        newnode = Node(data)
        if self.head is None:
            self.head = newnode
        else:
            current = self.head
            while current.next is not None:
                current = current.next
            current.next = newnode
        self.size += 1
        self.display()

    def insertAtIndex(self, data, index):
        if index == 0:
            self.insertAtBeginning(data)
            return
        if index == self.size:
            self.insertAtEnd(data)
            return
        newnode = Node(data)
        temp = self.head
        for i in range(1, index - 1):
            temp = temp.next
        newnode.next = temp.next
        temp.next = newnode
        self.size += 1
        self.display()

    def deleteFirst(self):
        if self.head is None:
            print("List already empty")
            return
        current = self.head
        self.head = self.head.next
        self.size -= 1
        print("Value Removed: ", current.data)
        return current.data

    def deleteLast(self):
        if self.size <= 1:
            return self.deleteFirst()
        secondLast = self.get(self.size - 2)
        val = self.get(self.size - 1).data
        secondLast.next = None
        self.size -= 1
        print("Value Removed: ", val)
        self.display()

    def deleteAtIndex(self, index):
        if index == 0:
            return self.deleteFirst()
        if index == self.size:
            return self.deleteLast()
        prev = self.get(index - 1)
        value = prev.next.data
        prev.next = prev.next.next
        self.size -= 1
        print("Value Removed: ", value)
        self.display()

    def get(self, index):
        newnode = self.head
        for i in range(index):
            newnode = newnode.next
        return newnode

    def display(self):
        if self.head is None:
            print("List is empty")
            print("The size is :", self.size)
            return
        current = self.head
        while current:
            print(current.data, end="->")
            current = current.next
        print("END")
        print("The size is :", self.size)


newLL = LL()
newLL.insertAtBeginning(10)
newLL.insertAtEnd(13)
newLL.insertAtEnd(12)
newLL.insertAtBeginning(3)
newLL.deleteFirst()
print("Value at index 2 is:", newLL.get(2).data)
newLL.insertAtIndex(31, 4)
newLL.insertAtEnd(43)
newLL.deleteLast()
newLL.deleteLast()
newLL.insertAtBeginning(89)
newLL.insertAtBeginning(34)
newLL.deleteAtIndex(3)
newLL.insertAtBeginning(45)
newLL.deleteAtIndex(1)
newLL.insertAtBeginning(67)
print("Final LL: ")
newLL.display()

File: single.cpp
#include<iostream>
using namespace std;

class Node  {
    public:
    int data;
    Node* next;
    Node(int data)  {
        this->data = data;
        this->next = nullptr;
    }
};


class LinkedList    {
    private:
    Node* head;
    int size;
    public:
    LinkedList()    {
        this->head = nullptr;
        this->size = 0;
    }

    void insertAtBeginning(int value)    {
        Node* newNode = new Node(value);
        newNode->next = head;
        head=newNode;
        size++;
        display();
    }

    void insertAtEnd(int value) {
        Node* newNode = new Node(value);
        if(head== nullptr) {
            head = newNode;
            size++;
            return;
        }
        Node* current = head;
        while(current->next != nullptr ){
            current = current->next;
        }
        current->next =newNode;
        size++;
        display();
    }

    void insertAtIndex(int value, int index)    {
        if(index==0)    {
            insertAtBeginning(value);
            return;
        }
        if(index==size) {
            insertAtEnd(value);
            return;
        }
        Node* temp = head;
        for(int i=1;i<index;i++)
        {   temp = temp->next;
        }
        Node* newNode = new Node(value);
        newNode->next = temp->next;
        temp->next = newNode;
        size++;
        display();
    }

    void deleteFirst()   {
        if(head== nullptr)  {
            cout<<"List is Empty";
            return;
        }
        Node* temp = head;
        head = temp->next;
        size--;
        cout<<"Value removed:"<<temp->data<<endl;
//        int value = head->data;
        delete temp;
        display();
    }

    void deleteLast()   {
        if(size<=1) {
            deleteFirst();
            return;
        }
        Node* temp = get(size-2);
        int value = get(size-1)->data;
        temp->next = nullptr;
        size--;
        cout<<"Value removed:"<<value<<endl;
        delete temp;
        display();
    }

    void deleteAtIndex(int index)   {
        if(index==0)    {
            deleteFirst();
            return;
        }
        if(index==size) {
            deleteLast();
            return;
        }
        Node* prev = get(index - 1);
        int value = prev->next->data;
        prev->next = prev->next->next;
        cout<<"Value removed:"<<value<<endl;
        size--;
        display();
//        delete temp;
    }

    Node* get(int index)    {
        Node* newnode = head;
        for(int i=0;i<index;i++)
        {   newnode =newnode->next;
        }
        return newnode;
    }

    void display()  {
        Node* current = head;
        if(head== nullptr)
        {   cout<<"List is Empty"<<endl;
            return ;
        }
        while(current!= nullptr)
        {   cout<<current->data<<"->";
            current = current->next;
        }
        cout<<"End"<<endl;
        cout<<"The size of list is: "<<size<<endl;
    }

};

int main()
{   LinkedList ll;
    ll.insertAtBeginning(12);
    ll.insertAtBeginning(11);
    ll.insertAtBeginning(10);
    cout<<ll.get(2)->data<<endl;
    ll.insertAtEnd(90);
    ll.insertAtIndex(31, 2);
    ll.insertAtBeginning(9);
    ll.deleteLast();
    ll.insertAtEnd(30);
    ll.insertAtIndex(25, 1);
    ll.insertAtBeginning(56);
    ll.deleteAtIndex(3);
    cout<<"Final LL: ";
    ll.display();
    return 0;
}